### PR TITLE
refactor: use flexible arrays instead of the length-of-one trick

### LINKED ISSUE #223: Buffer overflow with _FORTIFY_SOURCE=2 ("length of 1" trick)
OS: Arch Linux x64 current
Neovim version: current master (6eece58)
Neovim was installed via https://aur.archlinux.org/packages/neovim-git/

Stacktrace:

``` bash
nvim                   
*** buffer overflow detected ***: nvim terminated
======= Backtrace: =========
/usr/lib/libc.so.6(+0x731ff)[0x7fee74a051ff]
/usr/lib/libc.so.6(__fortify_fail+0x37)[0x7fee74a88157]
/usr/lib/libc.so.6(+0xf4380)[0x7fee74a86380]
nvim(eval_init+0xf4)[0x4498c4]
nvim(main+0x82)[0x422762]
/usr/lib/libc.so.6(__libc_start_main+0xf5)[0x7fee749b3b05]
nvim[0x4248d8]
======= Memory map: ========
00400000-005e8000 r-xp 00000000 fe:00 15889279                           /usr/bin/nvim
007e7000-007e8000 r--p 001e7000 fe:00 15889279                           /usr/bin/nvim
007e8000-007fb000 rw-p 001e8000 fe:00 15889279                           /usr/bin/nvim
007fb000-00806000 rw-p 00000000 00:00 0 
016fd000-01720000 rw-p 00000000 00:00 0                                  [heap]
7fee74158000-7fee7416d000 r-xp 00000000 fe:00 537168177                  /usr/lib/libgcc_s.so.1
7fee7416d000-7fee7436d000 ---p 00015000 fe:00 537168177                  /usr/lib/libgcc_s.so.1
7fee7436d000-7fee7436e000 rw-p 00015000 fe:00 537168177                  /usr/lib/libgcc_s.so.1
7fee7436e000-7fee74371000 r-xp 00000000 fe:00 537157256                  /usr/lib/libdl-2.19.so
7fee74371000-7fee74570000 ---p 00003000 fe:00 537157256                  /usr/lib/libdl-2.19.so
7fee74570000-7fee74571000 r--p 00002000 fe:00 537157256                  /usr/lib/libdl-2.19.so
7fee74571000-7fee74572000 rw-p 00003000 fe:00 537157256                  /usr/lib/libdl-2.19.so
7fee74572000-7fee74587000 r-xp 00000000 fe:00 537163252                  /usr/lib/libnsl-2.19.so
7fee74587000-7fee74786000 ---p 00015000 fe:00 537163252                  /usr/lib/libnsl-2.19.so
7fee74786000-7fee74787000 r--p 00014000 fe:00 537163252                  /usr/lib/libnsl-2.19.so
7fee74787000-7fee74788000 rw-p 00015000 fe:00 537163252                  /usr/lib/libnsl-2.19.so
7fee74788000-7fee7478a000 rw-p 00000000 00:00 0 
7fee7478a000-7fee74791000 r-xp 00000000 fe:00 537171620                  /usr/lib/librt-2.19.so
7fee74791000-7fee74990000 ---p 00007000 fe:00 537171620                  /usr/lib/librt-2.19.so
7fee74990000-7fee74991000 r--p 00006000 fe:00 537171620                  /usr/lib/librt-2.19.so
7fee74991000-7fee74992000 rw-p 00007000 fe:00 537171620                  /usr/lib/librt-2.19.so
7fee74992000-7fee74b30000 r-xp 00000000 fe:00 537171626                  /usr/lib/libc-2.19.so
7fee74b30000-7fee74d30000 ---p 0019e000 fe:00 537171626                  /usr/lib/libc-2.19.so
7fee74d30000-7fee74d34000 r--p 0019e000 fe:00 537171626                  /usr/lib/libc-2.19.so
7fee74d34000-7fee74d36000 rw-p 001a2000 fe:00 537171626                  /usr/lib/libc-2.19.so
7fee74d36000-7fee74d3a000 rw-p 00000000 00:00 0 
7fee74d3a000-7fee74d9a000 r-xp 00000000 fe:00 549028937                  /usr/lib/libncursesw.so.5.9
7fee74d9a000-7fee74f99000 ---p 00060000 fe:00 549028937                  /usr/lib/libncursesw.so.5.9
7fee74f99000-7fee74f9d000 r--p 0005f000 fe:00 549028937                  /usr/lib/libncursesw.so.5.9
7fee74f9d000-7fee74f9f000 rw-p 00063000 fe:00 549028937                  /usr/lib/libncursesw.so.5.9
7fee74f9f000-7fee74fb7000 r-xp 00000000 fe:00 537311998                  /usr/lib/libpthread-2.19.so
7fee74fb7000-7fee751b6000 ---p 00018000 fe:00 537311998                  /usr/lib/libpthread-2.19.so
7fee751b6000-7fee751b7000 r--p 00017000 fe:00 537311998                  /usr/lib/libpthread-2.19.so
7fee751b7000-7fee751b8000 rw-p 00018000 fe:00 537311998                  /usr/lib/libpthread-2.19.so
7fee751b8000-7fee751bc000 rw-p 00000000 00:00 0 
7fee751bc000-7fee751dc000 r-xp 00000000 fe:00 549330619                  /usr/lib/libuv.so.11.0.0
7fee751dc000-7fee753db000 ---p 00020000 fe:00 549330619                  /usr/lib/libuv.so.11.0.0
7fee753db000-7fee753dc000 r--p 0001f000 fe:00 549330619                  /usr/lib/libuv.so.11.0.0
7fee753dc000-7fee753dd000 rw-p 00020000 fe:00 549330619                  /usr/lib/libuv.so.11.0.0
7fee753dd000-7fee754dd000 r-xp 00000000 fe:00 537156455                  /usr/lib/libm-2.19.so
7fee754dd000-7fee756dc000 ---p 00100000 fe:00 537156455                  /usr/lib/libm-2.19.so
7fee756dc000-7fee756dd000 r--p 000ff000 fe:00 537156455                  /usr/lib/libm-2.19.so
7fee756dd000-7fee756de000 rw-p 00100000 fe:00 537156455                  /usr/lib/libm-2.19.so
7fee756de000-7fee756fe000 r-xp 00000000 fe:00 537171507                  /usr/lib/ld-2.19.so
7fee758b9000-7fee758ba000 rw-p 00000000 00:00 0 
7fee758ba000-7fee758bb000 rw-p 00000000 00:00 0 
7fee758bb000-7fee758bc000 rw-p 00000000 00:00 0 
7fee758bc000-7fee758bd000 rw-p 00000000 00:00 0 
7fee758bd000-7fee758be000 rw-p 00000000 00:00 0 
7fee758be000-7fee758bf000 rw-p 00000000 00:00 0 
7fee758fd000-7fee758fe000 rw-p 00000000 00:00 0 
7fee758fe000-7fee758ff000 r--p 00020000 fe:00 537171507                  /usr/lib/ld-2.19.so
7fee758ff000-7fee75900000 rw-p 00021000 fe:00 537171507                  /usr/lib/ld-2.19.so
7fee75900000-7fee75901000 rw-p 00000000 00:00 0 
7fffed14f000-7fffed170000 rw-p 00000000 00:00 0                          [stack]
7fffed1fe000-7fffed200000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
[1]    18073 abort (core dumped)  nvim

```

<bountysource-plugin>

---

Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/1429739-buffer-overflow-while-starting?utm_campaign=plugin&utm_content=tracker%2F461131&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F461131&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by dundargoc**
The "length-of-one" trick, where the last element of a struct is an
array of size 1, but extra size is allocated when calling malloc where
it uses more than 1 element in the array, cause problems with some
compilers. Some compilers set _FORTIFY_SOURCE=2 by default which
incorrectly considers it as an overflow. More information:
https://github.com/neovim/neovim/issues/223#issuecomment-1413828554

Using flexible array members allows us to to properly convey to the
compiler that its size may be larger than 1. This also enables us to
remove lengthy workarounds that are unreliable, as they depend on
CMAKE_BUILD_TYPE which isn't defined for multi-config generators.

Closes: https://github.com/neovim/neovim/issues/223

### CODE REVIEW

**Comment by zeertzjq** [comment]
Actually this is already done for `uf_name`:
https://github.com/neovim/neovim/commit/7da8056607331f12b912b4a7c65592ac317c2ff4

File: src/nvim/memline.c [hunk 1]
```diff
@@ -2720,7 +2720,7 @@ static bhdr_T *ml_new_ptr(memfile_T *mfp)
   PTR_BL *pp = hp->bh_data;
   pp->pb_id = PTR_ID;
   pp->pb_count = 0;
-  pp->pb_count_max = (uint16_t)((mfp->mf_page_size - sizeof(PTR_BL)) / sizeof(PTR_EN) + 1);
+  pp->pb_count_max = (uint16_t)((mfp->mf_page_size - sizeof(PTR_BL) - 1) / sizeof(PTR_EN) + 1);
```
Comment (zeertzjq) [inline src/nvim/memline.c:2723]: The size of `PTR_EN` is not 1

Comment (dundargoc) [reply src/nvim/memline.c]: Thanks.

**Comment by jamessan** [comment]
> * ci: add release job to CI

Shouldn't we be using at least `-O1` for the ASAN job anyway, since some of the issues that get flagged only occur with optimizations?

**Comment by dundargoc** [comment]
Sure, that sounds good. Changed ASAN to relwithdebinfo build.

File: src/nvim/ui_client.c [hunk 2]
```diff
@@ -59,7 +58,10 @@ uint64_t ui_client_start_server(int argc, char **argv)
 #ifdef MSWIN
     os_open_conin_fd();
 #else
-    dup(stderr_isatty ? STDERR_FILENO : STDOUT_FILENO);
+    int result = dup(stderr_isatty ? STDERR_FILENO : STDOUT_FILENO);
+    if (result == -1) {
+      exit(1);
```
Comment (zeertzjq) [inline src/nvim/ui_client.c:63]:
```suggestion
      return 0;
```

File: src/nvim/ui_client.c [hunk 3]
```diff
@@ -59,7 +58,10 @@ uint64_t ui_client_start_server(int argc, char **argv)
 #ifdef MSWIN
     os_open_conin_fd();
 #else
-    dup(stderr_isatty ? STDERR_FILENO : STDOUT_FILENO);
+    int result = dup(stderr_isatty ? STDERR_FILENO : STDOUT_FILENO);
```
Comment (bfredl) [inline src/nvim/ui_client.c:61]: can just use `(void)` to ignore this. if neither fd is usable we will never get here.

File: .github/workflows/env.sh [hunk 4]
```diff
@@ -22,14 +22,15 @@ EOF
...
-BUILD_FLAGS="CMAKE_FLAGS=-DCI_BUILD=ON -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX:PATH=$HOME/nvim-install -DBUSTED_OUTPUT_TYPE=nvim -DDEPS_PREFIX=$HOME/nvim-deps/usr -DMIN_LOG_LEVEL=3"
+BUILD_FLAGS="CMAKE_FLAGS=-DCI_BUILD=ON -DCMAKE_INSTALL_PREFIX:PATH=$HOME/nvim-install -DBUSTED_OUTPUT_TYPE=nvim -DDEPS_PREFIX=$HOME/nvim-deps/usr -DMIN_LOG_LEVEL=3"
 
 case "$FLAVOR" in
   asan)
     cat <<EOF >> "$GITHUB_ENV"
 CLANG_SANITIZER=ASAN_UBSAN
 ASAN_OPTIONS=detect_leaks=1:check_initialization_order=1:log_path=$GITHUB_WORKSPACE/build/log/asan:intercept_tls_get_addr=0
 UBSAN_OPTIONS=print_stacktrace=1 log_path=$GITHUB_WORKSPACE/build/log/ubsan
+BUILD_FLAGS="$BUILD_FLAGS -D CMAKE_BUILD_TYPE=RelWithDebInfo"
```
Comment (jamessan) [inline .github/workflows/env.sh:33]: This doesn't seem to be taking effect -- https://github.com/neovim/neovim/actions/runs/4073768577/jobs/7018093223#step:10:28

Comment (dundargoc) [reply .github/workflows/env.sh]: Thank you. Fixed.

**Comment by dundargoc** [comment]
I'll split the PR as the release CI is a red herring that isn't related anymore. I'll make this one about fortify_source/flexible array members only and create a new one for the release CI.

**Pull request merged**